### PR TITLE
KAFKA-4705: ConfigDef should support deprecated keys as synonyms

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 1]
```diff
@@ -99,10 +103,16 @@ public ConfigDef define(ConfigKey key) {
         if (configKeys.containsKey(key.name)) {
             throw new ConfigException("Configuration " + key.name + " is defined twice.");
         }
+        configKeys.put(key.name, key);
+        for (String name: key.deprecatedNames) {
+            if (configKeys.containsKey(name)) {
+                throw new ConfigException("Configuration " + key.name + " is defined twice.");
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Since it's tied to the deprecated name, the error message should probably include both the deprecated and new name. Otherwise it'll be more difficult to track down the source of the conflict.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 2]
```diff
@@ -424,19 +457,27 @@ public ConfigDef withClientSaslSupport() {
...
-            // props map contains setting - assign ConfigKey value
+            Object value = key.defaultValue;
             if (props.containsKey(key.name)) {
                 value = parseType(key.name, props.get(key.name), key.type);
-                // props map doesn't contain setting, the key is required because no default value specified - its an error
-            } else if (key.defaultValue == NO_DEFAULT_VALUE) {
-                throw new ConfigException("Missing required configuration \"" + key.name + "\" which has no default value.");
             } else {
-                // otherwise assign setting its default value
-                value = key.defaultValue;
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: It seems like we lost the case where we should set the value to the `key.defaultValue`. You more thoroughly handle other cases (deprecated values, no default value), but I think this line was removed and should still be at the tail end of the branches that have been added to this logic.

**Comment by ijuma** [comment]
There is a minor merge conflicts that needs to be resolved.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: ok

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 2 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: The object is initialized to `key.defaultValue`, so we don't need another clause at the end.  But it might be easier to understand with that at the end, so I'll re-add it.

**Comment by mjsax** [comment]
@cmccabe What is the status of this? I think it would be nice to get it for `0.10.2.1` -- not sure if you find time finishing this up. My PR #2648 depends on this one :)

**Comment by ewencp** [comment]
@cmccabe ping, looks like we might only need a rebase on this PR to get it merged

**Comment by cmccabe** [comment]
rebased

**Comment by guozhangwang** [comment]
LGTM.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5701/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5715/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by cmccabe** [comment]
Test failure is a Jenkins environment issue:
```
Total time: 1 hrs 27 mins 13.543 secs
[FINDBUGS] Collecting findbugs analysis files...
Setting GRADLE_3_4_RC_2_HOME=/home/jenkins/jenkins-slave/tools/hudson.plugins.gradle.GradleInstallation/Gradle_3.4-rc-2
ERROR: Step ?Publish FindBugs analysis results? aborted due to exception: 
java.lang.OutOfMemoryError: GC overhead limit exceeded
Caused: java.io.IOException: Remote call on ubuntu-2 failed
	at hudson.remoting.Channel.call(Channel.java:838)
	at hudson.FilePath.act(FilePath.java:985)
Caused: java.io.IOException: remote file operation failed: /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.12 at hudson.remoting.Channel@4cfb1c65:ubuntu-2
	at hudson.FilePath.act(FilePath.java:992)
	at hudson.FilePath.act(FilePath.java:974)
```

**Comment by cmccabe** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5725/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5711/
Test PASSed (JDK 8 and Scala 2.12).

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 3]
```diff
@@ -109,10 +114,17 @@ public ConfigDef define(ConfigKey key) {
         if (configKeys.containsKey(key.name)) {
             throw new ConfigException("Configuration " + key.name + " is defined twice.");
         }
+        configKeys.put(key.name, key);
+        for (String name: key.deprecatedNames) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Nit: space missing before `:`. There are other cases like this in the file.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 4]
```diff
@@ -447,10 +500,34 @@ public ConfigDef withClientSaslSupport() {
...
         Map<String, Object> values = new HashMap<>();
-        for (ConfigKey key : configKeys.values())
-            values.put(key.name, parseValue(key, props.get(key.name), props.containsKey(key.name)));
+        for (ConfigKey key : configKeys.values()) {
+            Object value = null;
+            if (props.containsKey(key.name)) {
+                value = parseType(key.name, props.get(key.name), key.type);
+            } else {
+                boolean foundDeprecated = false;
+                for (String name: key.deprecatedNames) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Nit: space before `:`.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 5]
```diff
@@ -447,10 +500,34 @@ public ConfigDef withClientSaslSupport() {
             String joined = Utils.join(undefinedConfigKeys, ",");
             throw new ConfigException("Some configurations in are referred in the dependents, but not defined: " + joined);
         }
-        // parse all known keys
+        // Handle all known ConfigKey objects.
         Map<String, Object> values = new HashMap<>();
-        for (ConfigKey key : configKeys.values())
-            values.put(key.name, parseValue(key, props.get(key.name), props.containsKey(key.name)));
+        for (ConfigKey key : configKeys.values()) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Does this mean that we process the same key multiple times if there are deprecated names (given https://github.com/apache/kafka/pull/2456/files#diff-0abcccc4b2e75cfa2ded051cf47076beR123)? Is that what we want?

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 6]
```diff
@@ -1094,6 +1176,19 @@ private void getConfigKeyRst(ConfigKey key, StringBuilder b) {
             b.append("  * Valid Values: ").append(getConfigValue(key, "Valid Values")).append("\n");
         }
         b.append("  * Importance: ").append(getConfigValue(key, "Importance")).append("\n");
+        if (key.deprecatedNames.size() > 0) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Nit: `!key.deprecatedNames.isEmpty`

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 7]
```diff
@@ -345,6 +382,20 @@ public ConfigDef define(String name, Type type, Importance importance, String do
...
     /**
+     * Define a new configuration with deprecated names but no special validation logic, no dependents and no custom recommender
+     * @param name              the name of the config parameter
+     * @param deprecatedNames   the deprecated names of the config parameter
+     * @param type              the type of the config
+     * @param importance        the importance of this config
+     * @param documentation     the documentation string for the config
+     * @return This ConfigDef so you can chain calls
+     */
+    public ConfigDef define(String name, Collection<String> deprecatedNames, Type type, Object defaultValue, Importance importance, String documentation) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: As you said offline, this is really getting out of hand. Should we introduce a builder of some sort? We can do it in a separate PR, but it would be good to do it before the next release so that we can remove the two overloads that have been introduced.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. Left a few questions and comments.

**Comment by mjsax** [comment]
@cmccabe Any update on this one?

**Comment by mjsax** [comment]
Ping an this again... It is still blocking one of my PRs... \cc @guozhangwang

**Comment by guozhangwang** [comment]
I already gave a +1 on this, and I think once @cmccabe addressed @ijuma 's comments it is good to go.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 7 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Good point.  I added a builder.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 5 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Good point... we don't need to process the same key twice.  I added a section to avoid this

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 8]
```diff
@@ -1188,6 +1338,14 @@ public void ensureValid(String name, Object value) {
         return updatedDependents;
     }
 
+    private static List<String> embeddedDeprecatedNames(final String keyPrefix, final List<String> deprecatedNames) {
+        final List<String> updatedDeprecatedNames = new ArrayList<>(deprecatedNames.size());
+        for (String dependent : deprecatedNames) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Nit: should it be `deprecated` instead of `dependent`?

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 9]
```diff
@@ -447,10 +552,36 @@ public ConfigDef withClientSaslSupport() {
...
+            Object value = null;
+            if (props.containsKey(key.name)) {
+                value = parseType(key.name, props.get(key.name), key.type);
+            } else {
+                boolean foundDeprecated = false;
+                for (String name : key.deprecatedNames) {
+                    if (props.containsKey(name)) {
+                        value = parseType(name, props.get(name), key.type);
+                        foundDeprecated = true;
+                        log.info("Configuration key {} is deprecated, and may be removed in the future.  " +
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Maybe we should say "and will be removed in a future major release". That makes it a bit less ambiguous.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 10]
```diff
@@ -447,10 +552,36 @@ public ConfigDef withClientSaslSupport() {
...
+                                "Please use {} instead.", name, key.name);
+                        break;
+                    }
+                }
+                if (!foundDeprecated) {
+                    if (key.defaultValue == NO_DEFAULT_VALUE)
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Nit: in `parseValue`, we changed this to `NO_DEFAULT_VALUE.equals(key.defaultValue)` due to findBugs warnings.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 11]
```diff
@@ -82,6 +87,96 @@
...
+        public ConfigDef build() {
+            ConfigKey configKey = new ConfigKey(name, type, defaultValue, validator, importance,
+                documentation, group, orderInGroup, width, displayName, dependents, recommender,
+                internalConfig, deprecatedNames);
+            return define(configKey);
+        }
+    }
+
+    Builder buildDefinition(String name, Type type, Importance importance) {
+        return new Builder(name, type, importance);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Should this be simply `builder`? The current name sounds like it builds the definition.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 12]
```diff
@@ -447,10 +552,36 @@ public ConfigDef withClientSaslSupport() {
             String joined = Utils.join(undefinedConfigKeys, ",");
             throw new ConfigException("Some configurations in are referred in the dependents, but not defined: " + joined);
         }
-        // parse all known keys
+        // Handle all known ConfigKey objects.
         Map<String, Object> values = new HashMap<>();
-        for (ConfigKey key : configKeys.values())
-            values.put(key.name, parseValue(key, props.get(key.name), props.containsKey(key.name)));
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:453]: We no longer use `parseValue` here, but `valuesWithPrefixOverride` still uses it. Will the latter handle deprecated key names correctly?

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 13]
```diff
@@ -447,10 +552,36 @@ public ConfigDef withClientSaslSupport() {
...
+        for (ConfigKey key : configKeys.values()) {
+            if (values.containsKey(key.name))
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Worth adding a comment explaining why this is needed. Also a comment in `configKeys` explaining that duplicate values may exist would be helpful.

**Review by ijuma — COMMENTED** [review]
Thanks for the update. A few more comments/questions.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 11 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: How about `configBuilder`

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 12 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:453]: That's a good point.  I fixed `valuesWithPrefixOverride` to work with deprecated config keys and added a unit test.  I also removed `parseValue`.

**Comment by mjsax** [comment]
@ijuma Can you have a look? This is still blocking one of my PRs.

**Comment by ijuma** [comment]
@mjsax it's near the top of my list.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java [hunk 14]
```diff
@@ -194,10 +194,13 @@ public Password getPassword(String key) {
...
         for (Map.Entry<String, ?> entry : originals.entrySet()) {
             if (entry.getKey().startsWith(prefix) && entry.getKey().length() > prefix.length()) {
-                String keyWithNoPrefix = entry.getKey().substring(prefix.length());
+                String prefixedName = entry.getKey();
+                String keyWithNoPrefix = prefixedName.substring(prefix.length());
                 ConfigDef.ConfigKey configKey = definition.configKeys().get(keyWithNoPrefix);
-                if (configKey != null)
-                    result.put(keyWithNoPrefix, definition.parseValue(configKey, entry.getValue(), true));
+                if (configKey != null) {
+                    result.put(configKey.name, ConfigDef.
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:201]: Why are we using the config key name as the key instead of `keyWithNoPrefix`?

Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:201]: `configKey.name` is always the true name, whereas `keyWithNoPrefix` may be a deprecated synonym.  We want the `values` map to contain true names as keys, not deprecated synonyms.

This is consistent with how the constructor of `AbstractConfig` invokes `ConfigDef#parse` to set  up `AbstractConfig#values`.  It should always map true names to values.  Its key set should not contain deprecated names.

Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:201]: The intention is that deprecated key strings are not referred to by anyone in the code ever-- except when defining `ConfigKey` objects.  All other parts of the code use the most current key names.

**Comment by mjsax** [comment]
Ping.

**Comment by mjsax** [comment]
Another regular ping here...

**Comment by mjsax** [comment]
Ping :)

**Comment by mjsax** [comment]
Retest this please.

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8069 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7961 tests run, 5 skipped, 2 failed.
--none--

**Comment by mjsax** [comment]
@cmccabe @ijuma What is the status here?

**Comment by ijuma** [comment]
Sorry, have been having a hard time getting back to this. Hopefully tomorrow.

**Comment by mjsax** [comment]
Periodic ping :)

**Comment by mjsax** [comment]
Can we get this for `1.1` ? \cc @ijuma @cmccabe

**Comment by mjsax** [comment]
Anything new here?

**Comment by cmccabe** [comment]
Closing as stale since this is quite old
